server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -204,14 +234,15 @@ public float getShardBalance() {
      * </ul>
      * <code>weight(node, index) = weight<sub>index</sub>(node, index) + weight<sub>node</sub>(node, index)</code>
      */
-    private static class WeightFunction {
+    public static class WeightFunction {
 
         private final float indexBalance;
         private final float shardBalance;
         private final float theta0;
         private final float theta1;
+        public AllocationConstraints constraints;
```

Comment:
Why public? You could potentially pass a lambda called extraWeight or something instead of making WeightFunction aware of AllocationConstraints.